overlord/devicestate/devicestate.go

```diff
@@ -462,298 +418,385 @@ func modelSnapChannelFromDefaultOrPinnedTrack(new *asserts.Model, s *asserts.Mod
 // pass both the snap name and the model snap, as it is possible that
 // the model snap is nil for UC16 models
 type modelSnapsForRemodel struct {
-	currentSnap            string
-	currentModelSnap       *asserts.ModelSnap
-	new                    *asserts.Model
-	newSnap                string
-	newModelSnap           *asserts.ModelSnap
-	newRequiredRevision    snap.Revision
-	newModelValidationSets *snapasserts.ValidationSets
-}
+	new *asserts.Model
+
+	oldSnap      string
+	oldModelSnap *asserts.ModelSnap
 
-func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
-	return ms.newModelSnap != nil &&
-		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
+	newSnap      string
+	newModelSnap *asserts.ModelSnap
 }
 
-type remodelVariant struct {
-	offline        bool
-	localSnaps     []*snap.SideInfo
-	localSnapPaths []string
+type remodeler struct {
+	newModel   *asserts.Model
+	offline    bool
+	localSnaps map[string]snapstate.PathSnap
+
+	// TODO:COMPS: keep track of local components here
+
+	vsets      *snapasserts.ValidationSets
+	tracker    *snap.SelfContainedSetPrereqTracker
+	deviceCtx  snapstate.DeviceContext
+	fromChange string
 }
 
-type pathSideInfo struct {
-	localSi *snap.SideInfo
-	path    string
+// remodelTarget represents a snap that is part of the model that we are
+// remodeling to.
+type remodelTarget struct {
+	// name is the name of the snap.
+	name string
+	// channel is the channel that the snap should be installed from and track.
+	channel string
+	// newModelSnap is the model snap for this target. This might be nil for
+	// either the snapd snap (which is implicitly in the model) or for the base
+	// snap on UC16 models. Always check for nil before using.
+	newModelSnap *asserts.ModelSnap
+	// oldModelSnap is the model snap for this target. This might be nil for
```

Comment:
this comment seems incomplete, should it mention something about previous model?